File: 1676-lowest-common-ancestor-of-a-binary-tree-iv/1676-lowest-common-ancestor-of-a-binary-tree-iv.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode[] nodes) {
        Set<TreeNode> set = new HashSet<>();
        for(TreeNode node : nodes) set.add(node);
        return LCA(root, set);
    }
    public TreeNode LCA(TreeNode root, Set<TreeNode> set) {
        //base case
        if(root == null) return null;
        if(set.contains(root)) return root;
        TreeNode left = LCA(root.left, set);
        TreeNode right = LCA(root.right, set);
        if(left != null && right != null) return root;
        else if(left != null) return left;
        else if(right != null) return right;
        else return null;
    }
}

File: 1676-lowest-common-ancestor-of-a-binary-tree-iv/1676-lowest-common-ancestor-of-a-binary-tree-iv.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', nodes: 'List[TreeNode]') -> 'TreeNode':
        targets = set()
        for node in nodes:
            targets.add(node)
        def dfs(root):
            if not root:
                return None
            if root in targets:
                return root
            left = dfs(root.left)
            right = dfs(root.right)
            if left and right:
                return root
            if not left and not right:
                return None
            return left or right
        return dfs(root)